Pointers training sections - Pull Request

DvdHalliday opened the pull request:
Disclaimer: The files that belong to the first section I just named as EX where X is the number of the exercise. The files of the second section are named S2EX.

Section 1:

Exercise 1
![E1](https://user-images.githubusercontent.com/109089386/187772710-0cdfefa2-6703-422f-8b56-31689a0a7bc2.png)
Exercise 2
![E2](https://user-images.githubusercontent.com/109089386/187772712-ec16bfae-dfc1-49fb-b325-873779b9cf45.png)
Exercise 3
![E3](https://user-images.githubusercontent.com/109089386/187772713-817a25e7-c137-46b4-900a-888e40be83c3.png)
Exercise  4
![E4](https://user-images.githubusercontent.com/109089386/187772718-40c1bd44-3fee-4c85-bb4c-c00d44d81173.png)
Exercise  5
![E5](https://user-images.githubusercontent.com/109089386/187772720-b35ae483-51b5-48d8-95b0-2d149bb6762b.png)
Exercise 6
![E6](https://user-images.githubusercontent.com/109089386/187772724-eeed05cc-4786-4451-aa40-93113aa0ae1a.png)
Exercise  7
![E7](https://user-images.githubusercontent.com/109089386/187772725-12ee52a7-0642-4e98-a87c-34e9c68a2d88.png)


Section 2:

Exercise 1
![E1](https://user-images.githubusercontent.com/109089386/187773085-2c012cef-4c5c-44e0-af2a-5e806b500c96.png)
Exercise  2
![E2](https://user-images.githubusercontent.com/109089386/187773087-e5b131ed-3d84-48a0-9b74-47e4298838dc.png)
Exercise 3
![E3](https://user-images.githubusercontent.com/109089386/187773090-807deda8-ee49-4acf-ba32-ab25ce2a685c.png)
Exercise 4
![E4](https://user-images.githubusercontent.com/109089386/187773092-1b970f56-feb4-4ec0-8539-05a5b0335098.png)
Exercise 5
![E5](https://user-images.githubusercontent.com/109089386/187773095-5393548c-db3b-477d-a8c2-a49386f83216.png)
Exercise 6
![E6](https://user-images.githubusercontent.com/109089386/187773096-49069e02-323f-413c-9db8-034220f85083.png)
Exercise 7
![E7](https://user-images.githubusercontent.com/109089386/187773097-e4439ce4-0e12-4d53-969f-0725c3638ca4.png)
Exercises 8 and 9 lack pics for they didn't actually change the output
Exercise 10
![E10](https://user-images.githubusercontent.com/109089386/187773099-d9033d00-6c83-46a6-83ee-8dde0605a724.png)
Exercise 11
![E11](https://user-images.githubusercontent.com/109089386/187773102-4e707166-3ae6-48af-9f3a-cfa66816204a.png)
Exercise 12
![E12](https://user-images.githubusercontent.com/109089386/187773106-9f008ae6-3bc8-4102-8fb9-89838dd5e965.png)**

DvdHalliday committed:
Exercise 1

Adding all the files that compose this first exercise

DvdHalliday committed:
Exercise 1 complement

Modified names of the project and files and also added the comment of the exercise statement.

DvdHalliday committed:
Exercise 2

Every fileee!

DvdHalliday committed:
Exercise 3

Summary: All the file you gon need

DvdHalliday committed:
Exercise 2 complement

Added the exercise title

DvdHalliday committed:
Exercise 3 complement

Added the exercise title

DvdHalliday committed:
Exercise 4

And it comes with every file you'd've dreamt

DvdHalliday committed:
Excercise 4 Complement

Small typo fix

DvdHalliday committed:
Exercise 5

For limited time only, every file you need.

DvdHalliday committed:
Exercise 6

& everythin' that comes with it

DvdHalliday committed:
Exercise 7

;3

DvdHalliday committed:
Section 2 Exercise 1

Simplicity

DvdHalliday committed:
Section 2 Exercise 2

I hope deleting all them 3 in a single line ain't not sin, mister

DvdHalliday committed:
Section 2 Exercise 3

Quiz done

DvdHalliday committed:
Section 2 Exercise 4

;3

DvdHalliday committed:
Section 2 Exercise 5

All the files.

DvdHalliday committed:
Section 2 Exercise 10

Every file including them headers!

DvdHalliday committed:
Section 2 Exercise 11

Concatenating the message I printed to console as an answer for the quiz of this exercise.

DvdHalliday committed:
Section 2 Exercise 12

Quiz complete

DvdHalliday committed:
Section 2 Exercise 6

DvdHalliday committed:
Section 2 Exercise 7

Temporary modifications.

DvdHalliday committed:
Section 2 Exercise 8

Destructor created ✓

DvdHalliday committed:
Section 2 Exercise 9

Before this I had the SetName method set as a private function, so that would be a way to prevent this memory leak from happening, however it makes way more sense to leave it as public in case someone wants to change it, and handle it as best as we can, to avoid memory leaks too.

DvdHalliday committed:
.sln upload

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E5/Source.cpp:
```diff
@@ -0,0 +1,108 @@
+	std::cout << "How many students do you want? ";
+
+	int NumberOfStudents;
+	std::cin >> NumberOfStudents;
```
I know that the exercise does not ask to verify the input but I think that it is best to continue with good programming practices by creating a function that could be called ValidIntInput(), since this is used more than once, which for me makes it useful in addition to making it difficult for the user to make mistakes

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E5/Source.cpp:
```diff
@@ -0,0 +1,108 @@
+
+		std::cout << "Enter the age of student #" << i + 1 << ": ";
+		int StudentAge;
+		std::cin >> StudentAge;
```
The same as the previous comment, it is solved with a function

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E5/Source.cpp:
```diff
@@ -0,0 +1,108 @@
+
+		std::cout << "Enter a name for student #" << i + 1 << ": ";
+		char TemporaryStudentName[32];
+		std::cin.getline(TemporaryStudentName, 32, '\n');
```
you need to validate this input, but it is not necessary to apply a function since it is used only in this place

camodesigns commented on c++ training/PointersTraining/PointerExercises/E2/Source.cpp:
```diff
@@ -0,0 +1,28 @@
+	std::cout << "Please enter a value for variable A: ";
+
+	int A;
+	std::cin >> A;
```
You need to validate entries, not because it is a requirement but to maintain the good practices that we have been using

camodesigns commented on c++ training/PointersTraining/PointerExercises/E2/Source.cpp:
```diff
@@ -0,0 +1,28 @@
+	std::cout << "Please enter a new value for variable A: ";
+
+	int* APointer = &A;
+	std::cin >> *APointer;
```
Again Validate input  please, we don't want surprises and less if it is with a pointer

camodesigns commented on c++ training/PointersTraining/PointerExercises/E3/Source.cpp:
```diff
@@ -0,0 +1,35 @@
+	std::cout << "Please enter a value for variable A: ";
+
+	int A;
+	std::cin >> A;
```
Validate Input whit a function, it's only a recomendation

camodesigns commented on c++ training/PointersTraining/PointerExercises/E3/Source.cpp:
```diff
@@ -0,0 +1,35 @@
+	int B;
+	std::cin >> B;
+
+	int* P = &A;
```
the letter p is a general standard for programmers, however for readability you should perhaps use another way to shorten the pointer, if that is what you are looking for, with names like Ptr that may be clearer for other programmers reading your code

DvdHalliday replied:
The exercise explicitly asked for us to use the P name for such variable though

camodesigns commented on c++ training/PointersTraining/PointerExercises/E7/Source.cpp:
```diff
@@ -0,0 +1,33 @@
+int main()
+{
+	int Three = 3;
+	int* ThreePointer = nullptr;
```
The name of this pointer can be more clear

DvdHalliday replied:
Why? If my variable is called three, what could be clearer than ThreePointer for a pointer that is pointing to it?

camodesigns commented on c++ training/PointersTraining/PointerExercises/Project1/Source.cpp:
```diff
@@ -0,0 +1,35 @@
+	std::cout << "Please enter a number: ";
+
+	int IntVariable;
+	std::cin >> IntVariable;
```
It can be annoying, but it is better to get used to using validation and inputs always

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E3/Source.cpp:
```diff
@@ -0,0 +1,52 @@
+
+
+	std::cout << "The first memory leak is simpler, you're creating an integer in heap memory, and once the function scope's over" << std::endl;
+	std::cout << "the variable will be lost for you to access, and so as well for deletion." << std::endl << std::endl;
```
You should also mention that this variable will take up space until you stop executing the program.

DvdHalliday replied:
Isn't that already explicit when we call it a memory leak?

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E3/Source.cpp:
```diff
@@ -0,0 +1,52 @@
+	std::cout << "when we define a function as such, all we are doing really, is creating a copy of the pointer," << std::endl;
+	std::cout << "a copy that is given an adress in the heap memory, so once again, whenever the function scope" << std::endl;
+	std::cout << "comes to an end, so will the pointer copy, hence the information of memory that will never" << std::endl;
+	std::cout << "be used again, will now be lost for us to delete." << std::endl;
```
In addition to not being able to use the pointer again, it will remain pointing to a garbage memory address indefinitely.

DvdHalliday replied:
That is not what I said, I said we won't be able to use memory again, the pointer could be easiliy used again.
Also it is not at all true that it will remain pointing to any garbage address, the first pointer we created, IntPtr, was pointing to null, and it will still be doing that, because what we're doing inside the function is modifying a copy of that original pointer, and this new copy will be destroy once the function scope comes to an end, so if you go check it in memory, IntPtr will still be pointing null at the end of the program, what's lost is the memory address, hence a memory leak.

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E4/Source.cpp:
```diff
@@ -0,0 +1,33 @@
+
+	if (SoonToBeADanglingPointer)
+	{
+		std::cout << *SoonToBeADanglingPointer;
```
I know that you print Soon To Be A Dangling Pointer, but it is not clear to me, it is not the expected validation, also in case the validation is not true, you are leaving it to the air and the idea is that you can try or test what happens with the exercise

DvdHalliday replied:
What do you mean it's not the expected validation?

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E4/Source.cpp:
```diff
@@ -0,0 +1,33 @@
+		std::cout << *SoonToBeADanglingPointer;
+	}
+
+	std::cout << "Well, this is simply happening cause even though the memory was set free again," << std::endl;
```
This is not entirely true, if you run the program it will appear that it is valid, because although the value is deleted, the memory address still exists, it probably points to a garbage value or a strange value, but the fact that it points to something that is not 0 or null makes us return that it is valid. For this very reason it is a risk to have this type of situation

DvdHalliday replied:
I feel like that's what I said, I don't see what you say is not entirely true.

DvdHalliday replied:
In case this is because it has no trouble anymore, it's because I decided to upload the version modified with this exercise recommendation that we should set it to nullptr once we get rid of it.

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E2/Source.cpp:
```diff
@@ -0,0 +1,43 @@
+
+	delete HeapInt1, HeapInt2, HeapInt3;
+
+	std::cout << "The difference between these memory addresses is: essentially, that the variables that areinitialised and allocated in the stack are " << std::endl;
```
please take care whit the spaces in that output areintialised != are initialised

camodesigns commented on c++ training/PointersTraining/PointerExercises/S2E2/Source.cpp:
```diff
@@ -0,0 +1,43 @@
+	std::cout << "being stored continuously one after another, every 4 bytes, which is the size of every integer." << std::endl;
+	std::cout << "Heap allocated integers are, on the other hand, given memory locations that seem much more random, there is in fact not necessarilly " << std::endl;
+	std::cout << "always much correlation between the variables addresses altogether. They might be near each other, however there will not always be" << std::endl;
+	std::cout << "the same distance between one another." << std::endl;
```
I can add that the difference is also that the first ones are stored on the stack and the others on the heap, which translates into a difference in data reading speed, in addition to being more than random (because pro yes both have random addresses in memory) is that in one the bytes are separated and it is known that the values will be or will be and in the other it is a promise, that you fulfill it or not is not a matter of dynamic memory

DvdHalliday replied:
I feel like the fact that some are stored or allocated in heap memory and stack memory was many times addressed in the answer.

DvdHalliday replied:
I'm not quite sure of what you're saying there, I haven't found whatsoever yet any information on the speed of data reading would change between these two. There is a difference, however, when you're actually asking to find free memory there is in fact a difference in speed, but only to access free memory, once you have it it's a matter of going directly to that address you now possess, so there should not be any difference.

DvdHalliday replied:
What I intend to show is that there is a clear difference in the separation between the two types of memory allocation, one of the is stored continuously while that other doesn't strictly follow such a clear path.

DvdHalliday committed:
Camilo's review changes

added an extra space in S2E2. I also added an extra behaviour for the S2E4 I hadn't actually done as the exercises asked for.
